File: api/conversation.py
import uuid
from datetime import datetime, timezone

from google.cloud import firestore


def create_conversation(db, uid: str, governing_body: str | None, preview: str) -> str:
    conv_id = str(uuid.uuid4())
    db.collection("users").document(uid).collection("conversations").document(conv_id).set({
        "created_at": datetime.now(timezone.utc),
        "governing_body": governing_body,
        "preview": preview[:100],
    })
    return conv_id


def append_message(
    db,
    uid: str,
    conversation_id: str,
    role: str,
    content: str,
    sources: list[dict] | None = None,
) -> None:
    data: dict = {
        "role": role,
        "content": content,
        "created_at": datetime.now(timezone.utc),
    }
    if sources is not None:
        data["sources"] = sources
    db.collection("users").document(uid).collection("conversations").document(
        conversation_id
    ).collection("messages").add(data)


def list_conversations(db, uid: str, limit: int = 50) -> list[dict]:
    docs = (
        db.collection("users")
        .document(uid)
        .collection("conversations")
        .order_by("created_at", direction=firestore.Query.DESCENDING)
        .limit(limit)
        .stream()
    )
    result = []
    for doc in docs:
        data = doc.to_dict()
        data["id"] = doc.id
        result.append(data)
    return result


def get_messages(db, uid: str, conversation_id: str) -> list[dict]:
    docs = (
        db.collection("users")
        .document(uid)
        .collection("conversations")
        .document(conversation_id)
        .collection("messages")
        .order_by("created_at")
        .stream()
    )
    result = []
    for doc in docs:
        data = doc.to_dict()
        data["id"] = doc.id
        result.append(data)
    return result

File: app/lib/models/conversation.dart
import 'message.dart';

class Conversation {
  const Conversation({
    required this.id,
    required this.preview,
    required this.createdAt,
    this.governingBody,
    this.messages = const [],
  });

  final String id;
  final String preview;
  final DateTime createdAt;
  final String? governingBody;
  final List<ConversationMessage> messages;

  factory Conversation.fromJson(Map<String, dynamic> json) => Conversation(
        id: json['id'] as String,
        preview: json['preview'] as String,
        createdAt: DateTime.parse(json['created_at'] as String),
        governingBody: json['governing_body'] as String?,
        messages: json['messages'] != null
            ? (json['messages'] as List)
                .map((m) => ConversationMessage.fromJson(m as Map<String, dynamic>))
                .toList()
            : const [],
      );
}

class ConversationMessage {
  const ConversationMessage({
    required this.id,
    required this.role,
    required this.content,
    required this.createdAt,
    this.sources = const [],
  });

  final String id;
  final String role;
  final String content;
  final DateTime createdAt;
  final List<Source> sources;

  factory ConversationMessage.fromJson(Map<String, dynamic> json) =>
      ConversationMessage(
        id: json['id'] as String,
        role: json['role'] as String,
        content: json['content'] as String,
        createdAt: DateTime.parse(json['created_at'] as String),
        sources: json['sources'] != null
            ? (json['sources'] as List)
                .map((s) => Source.fromJson(s as Map<String, dynamic>))
                .toList()
            : const [],
      );
}
